subprojects/core/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -0,0 +1,399 @@
+            } finally {
+                withoutWorkerLease(new Runnable() {
+                    @Override
+                    public void run() {
+                        for (DefaultProjectLock projectLock : threadProjectMap.get(threadId)) {
+                            projectLock.lock();
+                        }
+                    }
+                });
+            }
+        } else {
+            runnable.run();
+        }
+    }
+
+    private class DefaultProjectLock implements ProjectLock {
+        private final String projectPath;
+        private ReentrantLock lock = new ReentrantLock();
+
+        DefaultProjectLock(String projectPath) {
+            this.projectPath = projectPath;
+        }
+
+        boolean tryLock() {
+            if (!lock.isHeldByCurrentThread()) {
+                if (lock.tryLock()) {
+                    LOGGER.debug("{}: acquired project lock on {}", Thread.currentThread().getName(), projectPath);
+                    return true;
+                } else {
+                    return false;
+                }
+            } else {
+                return true;
+            }
+        }
+
+        void unlock() {
+            if (lock.isHeldByCurrentThread()) {
+                lock.unlock();
+                LOGGER.debug("{}: released project lock on {}", Thread.currentThread().getName(), projectPath);
+                projectLockBroadcast.onProjectUnlock(projectPath);
+            }
+        }
+
+        void lock() {
+            if (!lock.isHeldByCurrentThread()) {
+                lock.lock();
+                LOGGER.debug("{}: acquired project lock on {}", Thread.currentThread().getName(), projectPath);
+            }
+        }
+
+        @Override
+        public boolean hasProjectLock() {
+            return lock.isHeldByCurrentThread();
+        }
+
+        @Override
+        public boolean isLocked() {
+            return lock.isLocked();
+        }
+
+        @Override
+        public boolean tryWithProjectLock(Runnable runnable) {
+            Long threadId = Thread.currentThread().getId();
+            final DefaultProjectLock thisLock = this;
+            if (!hasProjectLock()) {
+                if (tryLock()) {
+                    try {
+                        threadProjectMap.put(threadId, thisLock);
+                        runnable.run();
+                    } finally {
+                        unlock();
+                        threadProjectMap.remove(threadId, thisLock);
+                    }
+                    return true;
+                }
+            } else {
+                runnable.run();
+                return true;
+            }
+
+            return false;
+        }
+
+        @Override
+        public void withProjectLock(Runnable runnable) {
+            final Long threadId = Thread.currentThread().getId();
+            final DefaultProjectLock thisLock = this;
+            if (!hasProjectLock()) {
+                withoutWorkerLease(new Runnable() {
```

Comment:
I've changed this so that we first try to reacquire any project locks without releasing the worker leases (non-blocking).  If we can't, then we release the worker leases and block until we can get all of the project locks and worker leases atomically.